src/python-common/ceph/rgw/rgwam_core.py

```diff
@@ -612,99 +676,128 @@ def realm_rm_zone_creds(self, realm_token_b64):
+        return (0, success_message, '')
 
-        # if endpoints:
-        #    eps = endpoints.split(',')
-        #    ep = ''
-        #    if len(eps) > 0:
-        #        ep = eps[0]
-        #        o = urlparse(ep)
-        #        port = o.port
-        #        spec = RGWSpec(service_id = svc_id,
-        #                       rgw_realm = realm_name,
-        #                       rgw_zone = zone_name,
-        #                       rgw_frontend_port = o.port)
-        #        self.env.mgr.apply_rgw(spec)
+    def get_realms_info(self):
+        realms_info = []
+        for realm_name in self.realm_op().list():
+            realm = self.get_realm(realm_name)
+            master_zone_inf = self.period_op().get_master_zone(realm)
+            zone_ep = self.period_op().get_master_zone_ep(realm)
+            if master_zone_inf and 'system_key' in master_zone_inf:
+                access_key = master_zone_inf['system_key']['access_key']
+                secret = master_zone_inf['system_key']['secret_key']
+            else:
+                access_key = ''
+                secret = ''
+            realms_info.append({"realm_name": realm_name,
+                                "realm_id": realm.id,
+                                "master_zone_id": master_zone_inf['id'] if master_zone_inf else '',
+                                "endpoint": zone_ep[0] if zone_ep else None,
+                                "access_key": access_key,
+                                "secret": secret})
+        return realms_info
+
+    def zone_create(self, rgw_spec, start_radosgw):
+
+        if not rgw_spec.rgw_realm_token:
+            raise RGWAMException('missing realm token')
+        if rgw_spec.rgw_zone is None:
+            raise RGWAMException('Zone name is a mandatory parameter')
+        if rgw_spec.rgw_zone in self.zone_op().list():
+            raise RGWAMException(f'Zone {rgw_spec.rgw_zone} already exists')
+
+        realm_token = RealmToken.from_base64_str(rgw_spec.rgw_realm_token)
+        if realm_token.endpoint is None:
+            raise RGWAMException('Provided realm token has no endpoint')
+
+        access_key = realm_token.access_key
+        secret = realm_token.secret
+        try:
+            realm_info = self.realm_op().pull(EntityName(realm_token.realm_name),
+                                              realm_token.endpoint, access_key, secret)
+        except RGWAMException as e:
+            raise RGWAMException('failed to pull realm', e)
 
-        self.env.mgr.apply_rgw(svc_id, realm_name, zone_name)
+        logging.info(f"Pulled realm {realm_info['name']} ({realm_info['id']})")
+        realm_name = realm_info['name']
+        realm_id = realm_info['id']
 
-        daemons = self.env.mgr.list_daemons(svc_id, 'rgw', refresh=True)
+        realm = EntityID(realm_id)
+        period_info = self.period_op().get(realm)
+        period = RGWPeriod(period_info)
+        logging.info('Period: ' + period.id)
```

Comment:
should this be logged at debug level? Would we need this period id for anything?

Reply:
Changed to debug. I guess it may be useful for debugging. It was there in the original code so I left the log message.